flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionType.java

```diff
@@ -81,7 +81,19 @@ public enum ResultPartitionType {
      * in that {@link #PIPELINED_APPROXIMATE} partition can be reconnected after down stream task
      * fails.
      */
-    PIPELINED_APPROXIMATE(true, false, ConsumingConstraint.CAN_BE_PIPELINED, ReleaseBy.UPSTREAM);
+    PIPELINED_APPROXIMATE(true, false, ConsumingConstraint.CAN_BE_PIPELINED, ReleaseBy.UPSTREAM),
+
+    /**
+     * Hybrid partitions with a bounded (local) buffer pool to support downstream task to
+     * simultaneous reading and writing shuffle data.
+     *
+     * <p>Hybrid result has the following two characteristics:
+     *
+     * <p>Intermediate data can be consumed any time, whether fully produced or not.
+     *
+     * <p>Intermediate data can be consumed directly from memory as much as possible.
+     */
+    HYBRID(true, false, ConsumingConstraint.CAN_BE_PIPELINED, ReleaseBy.SCHEDULER);
```

Comment:
Is this type will be included in the `isBlockingOrBlockingPersistentResultPartition` or `isPipelinedOrPipelinedBoundedResultPartition` ? 

I found that after last change, these two new methods functionality is not very clear, It works like hard code check the `Pipelined` and `Blocking` type, Right?

Reply:
This will not be included in `isBlockingOrBlockingPersistentResultPartition` or `isPipelinedOrPipelinedBoundedResultPartition`.

The two methods here are exactly, as you said, checking for the specific types of the interface implementations. This is obviously not a good design, because you should not assume working with a specific implementation of the interface. However, this was not newly introduced. FLINK-27902 only explicitly separates use cases that rely on specific implementations from those that properly rely on the interfaces. Fixing of them probably requires more careful redesign in the problematic use cases, and we do not want to block this feature on that.

Reply:
Yes, the interfaces were intended to expose a characteristic rather than to be bounded to a specific implementation. The problem was on the caller side. Instead of relying on the intended characteristic, some callers were relying on assumptions such as "a result partition type that returns `true` for `isBlocking` must be `PIPELINED` or `PIPELINED_BOUNDED`".

E.g., in `AdaptiveScheduler#assertPreconditions`, the intention here is to make sure only `PIPELINED` and `PIPELINED_BOUNDED` are involved. You may take a look at where `isBlockingOrBlockingPersistentResultPartition` and `isPipelinedOrPipelinedBoundedResultPartition` for more details.

Reply:
Get it, thanks